### PR TITLE
Consul

### PR DESCRIPTION
**Pull request by Malyna23**
Add consul ssl

### CODE REVIEW

File: modules/consul/instance.tf [hunk 1]
```diff
@@ -29,9 +29,17 @@ resource "null_resource" consul_cluster {
   }
 
   provisioner "file" {
-    source      = "keys"
+    source      = "keys/${count.index}"
     destination = "/home/ubuntu/keys"
   }
+   provisioner "file" {
+    source      = "keys/ca.pem"
```
Comment (imknysh) [inline modules/consul/instance.tf:38]: try to use https://www.terraform.io/docs/providers/template/d/file.html and don't store keys locally

File: modules/databases/instance.tf [hunk 2]
```diff
@@ -0,0 +1,11 @@
+resource "aws_db_instance" "default" {
+  allocated_storage    = 20
+  storage_type         = "gp2"
+  engine               = "mysql"
+  engine_version       = "5.7"
+  instance_class       = "db.t2.micro"
+  name                 = ""
```
Comment (imknysh) [inline modules/databases/instance.tf:16]: Try to use variables for all parameters that could be changed in the future

**Comment by imknysh** [comment]
Please add gossip encryption https://www.consul.io/docs/agent/encryption.html

File: provider.tf [hunk 3]
```diff
@@ -0,0 +1,13 @@
+terraform {
+  backend "s3" {
+    bucket                  = "consultfstate"
+    region                  = "eu-central-1"
+    key                     = "terraform.tfstate"
+    shared_credentials_file = "./credentials"
```
Comment (vovandodev) [inline provider.tf]: Don't use explicit credentials defining. It's an optional parameter and your keys should be stored in the default for terraformi location.  Your code in the future will be used by different members so they shouldn't specify anything custom.

File: provider.tf [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  }
+}
+
+provider "aws" {
+  region                  = "${var.region}"
+  shared_credentials_file = "./credentials"
```
Comment (vovandodev) [inline provider.tf]: the same here. useless

File: modules/databases/instance.tf [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+resource "aws_db_instance" "default" {
+  allocated_storage    = 20
+  storage_type         = "gp2"
+  engine               = "mysql"
+  engine_version       = "5.7"
+  instance_class       = "db.t2.micro"
+  name                 = ""
+  username             = ""
```
Comment (vovandodev) [inline modules/databases/instance.tf:17]: If parameters are optional and you don't specify them - don't add them. It's useless and makes your code hardly readable.

File: modules/consul/variables.tf [hunk 6]
```diff
@@ -0,0 +1,7 @@
+variable "count_consul" { default = "2" }
+variable "ami" { default = "ami-090f10efc254eaf55" }
+variable "instance_type" { default = "t2.micro" }
+variable "availability_zone" { default = "eu-central-1a" }
+variable "vpc_network" {default = []}
+variable "security_group" {}
+variable "aws_pem_key_file_path" { default = "./modules/consul/.ssh/id_rsa" }
```
Comment (vovandodev) [inline modules/consul/variables.tf:7]: Do not hardcode ssh key path. Use ssh-agent with keys instead: add your key and be happy.

File: modules/consul/instance.tf (hunk 1 above)
Comment (vovandodev) [reply modules/consul/instance.tf:38]: He tried to do it below. How do you see this approach especially here?

File: modules/consul/instance.tf [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    source      = "modules/consul/templates/config.json"
+    destination = "/home/ubuntu/config.json"
+  }
+
+  provisioner "file" {
+    content     = "${element(data.template_file.vault_conf.*.rendered, count.index)}"
+    destination = "/home/ubuntu/vault/config/local.json"
+  }
+
+  provisioner "remote-exec" {
```
Comment (vovandodev) [inline modules/consul/instance.tf:57]: In this particular case, it's better to use templates and conduct provision script in a separate file. In Production better to use SCM.

File: main.tf [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+   certs_count = "${var.count_consul}"
+   private_key_pem = "${module.ca.ca_private_key_pem}"
+   ca_cert_pem  = "${module.ca.ca_cert_pem}"
+   private_dns = "${module.consul.private_dns}"
+   private_ip_all = "${module.consul.private_ip_all}"
+
+ }
+ module "ca" {
+   source = "./modules/ca"
+   version = "0.0.1"
```
Comment (vovandodev) [inline main.tf]: Why do you need this here?

File: main.tf [hunk 9]
```diff
@@ -0,0 +1,31 @@
+ module "consul" {
+   source = "./modules/consul"
+   count_consul = "${var.count_consul}"
+   version = "0.0.1"
```
Comment (vovandodev) [inline main.tf]: How do you handle versioning? As I remember this parameter is only for modules installed from a module registry, such as the Terraform Registry or Terraform Enterprise's private module registry.

**Comment by vovandodev** [comment]
Use modules only when you really need it. You have a lot of useless files and code. :)

**Pull request merged**